src/tests/test_family_tree.c

```diff
@@ -0,0 +1,446 @@
+
+cgsize_t sizes[9], range[6];
+
+float xcoord[NUM_SIDE*NUM_SIDE*NUM_SIDE];
+float ycoord[NUM_SIDE*NUM_SIDE*NUM_SIDE];
+float zcoord[NUM_SIDE*NUM_SIDE*NUM_SIDE];
+
+void error_exit (char *where)
+{
+    fprintf (stderr, "ERROR:%s:%s\n", where, cg_get_error());
+    exit (1);
+}
+
+int main ()
+{
+    int i, j, k, n, nfam, nb, ng, nnames, ier;
+
+    int cgfile, cgbase, cgtree, cgzone, cgfam, cgcoord, cgbc, cgsr;
+    float exp[5];
+    char outfile[33], name[33], tname[33];
+    char family_name[20*33];
+    char tfamily_name[20*33];
+
+    /* ================================================================ WRITING TESTS */
+
+    strcpy( outfile, "family_tree.cgns" );
+    unlink( outfile );
+
+    TRACE( "Create file" );
+    if( cg_open( outfile, CG_MODE_WRITE, &cgfile ) )
+        error_exit( "cg_open" );
+
+    TRACE( "Create Grid Base" );
+    if( cg_base_write( cgfile, "Structured", 3, 3, &cgbase ) ||
+        cg_gopath( cgfile, "/Structured" )                   ||
+        cg_dataclass_write( CGNS_ENUMV(Dimensional))         ||
+        cg_units_write( CGNS_ENUMV(Kilogram),
+                        CGNS_ENUMV(Meter),
+                        CGNS_ENUMV(Second),
+                        CGNS_ENUMV(Kelvin),
+                        CGNS_ENUMV(Radian) ) )
+        error_exit( "grid base write" );
+
+    TRACE( "Create Family Tree Base" );
+    if( cg_base_write( cgfile, "FamilyTree", 3, 3, &cgtree ) )
+        error_exit( "family tree base");
+
+    /* SOME GRID DATA */
+    for (n = 0; n < 3; n++) {
+        sizes[n]   = NUM_SIDE;
+        sizes[n+3] = NUM_SIDE - 1;
+        sizes[n+6] = 0;
+    }
+    for (n = 0, k = 0; k < NUM_SIDE; k++) {
+        for (j = 0; j < NUM_SIDE; j++) {
+            for (i = 0; i < NUM_SIDE; i++, n++) {
+                xcoord[n] = (float)i;
+                ycoord[n] = (float)j;
+                zcoord[n] = (float)k;
+            }
+        }
+    }
+
+    for (n = 0; n < 5; n++)
+        exp[n] = (float)0;
+    exp[1] = (float)1;
+    if (cg_zone_write(cgfile, cgbase, "Zone", sizes,
+            CGNS_ENUMV(Structured), &cgzone))
+        error_exit("zone write");
+    if (cg_coord_write(cgfile, cgbase, cgzone, CGNS_ENUMV(RealSingle),
+            "CoordinateX", xcoord, &cgcoord) ||
+        cg_coord_write(cgfile, cgbase, cgzone, CGNS_ENUMV(RealSingle),
+            "CoordinateY", ycoord, &cgcoord) ||
+        cg_coord_write(cgfile, cgbase, cgzone, CGNS_ENUMV(RealSingle),
+            "CoordinateZ", zcoord, &cgcoord) ||
+        cg_goto(cgfile, cgbase, "Zone_t", cgzone, "GridCoordinates", 0,
+            "CoordinateX", 0, NULL) ||
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp) ||
+        cg_gopath(cgfile, "../CoordinateY") ||
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp) ||
+        cg_gopath(cgfile, "../CoordinateZ") ||
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp))
+        error_exit("coordinate write");
+
+    /* Write Family Tree */
+    /*
+     * FamilyTree
```

Comment:
Probably we need a limits test with (1) empty strings  (2) only / chars (3) max number of chars

Reply:
Should we add:
empty string --> check that it fails
only '/' --> check that it fails
max _number_of_char  --> pass
max _number_of_char +1 --> fail
??

Reply:
I added the tests